File: code/c/08-dfs_search.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define MAX_NODES 50000  // Adjust as needed

/**
 * Node structure for adjacency list
 */
typedef struct Node {
  int vertex;
  struct Node* next;
} Node;

/**
 * Graph structure
 */
typedef struct Graph {
  int numVertices;
  Node** adjLists;
  int* visited;
} Graph;

/**
 * Function to create a node
 */
Node* createNode(int v) {
  Node* newNode = (Node*)malloc(sizeof(Node));
  newNode->vertex = v;
  newNode->next = NULL;
  return newNode;
}

/**
 * Function to create a graph
 */
Graph* createGraph(int vertices) {
  Graph* graph = (Graph*)malloc(sizeof(Graph));
  graph->numVertices = vertices;

  graph->adjLists = (Node**)malloc(vertices * sizeof(Node*));
  graph->visited = (int*)malloc(vertices * sizeof(int));

  for (int i = 0; i < vertices; i++) {
    graph->adjLists[i] = NULL;
    graph->visited[i] = 0;
  }

  return graph;
}

/**
 * Function to add edge to an undirected graph
 */
void addEdge(Graph* graph, int src, int dest) {
  // Add edge from src to dest
  Node* newNode = createNode(dest);
  newNode->next = graph->adjLists[src];
  graph->adjLists[src] = newNode;

  // Since the graph is undirected, add edge from dest to src also
  newNode = createNode(src);
  newNode->next = graph->adjLists[dest];
  graph->adjLists[dest] = newNode;
}

/**
 * DFS algorithm
 */
void dfs(Graph* graph, int vertex) {
  graph->visited[vertex] = 1;

  Node* temp = graph->adjLists[vertex];

  while (temp) {
    int adjVertex = temp->vertex;

    if (graph->visited[adjVertex] == 0) {
      dfs(graph, adjVertex);
    }
    temp = temp->next;
  }
}

/**
 * Main function to demonstrate DFS and measure execution time.
 */
int main() {
  int N_values[] = {1000, 5000, 10000, 20000, 50000};
  int num_tests = sizeof(N_values) / sizeof(N_values[0]);

  for (int t = 0; t < num_tests; t++) {
    int N = N_values[t];
    if (N > MAX_NODES) {
      printf("N exceeds MAX_NODES limit.\n");
      continue;
    }

    Graph* graph = createGraph(N);

    // Build a connected graph (e.g., a chain)
    for (int i = 0; i < N - 1; i++) {
      addEdge(graph, i, i + 1);
    }

    // Measure start time
    clock_t start_time = clock();

    // Perform DFS starting from vertex 0
    dfs(graph, 0);

    // Measure end time
    clock_t end_time = clock();

    // Calculate time taken
    double time_spent = (double)(end_time - start_time) / CLOCKS_PER_SEC;

    // Print the result
    printf("DFS with N = %d\n", N);
    printf("Time taken: %f seconds\n\n", time_spent);

    // Free memory
    for (int i = 0; i < N; i++) {
      Node* temp = graph->adjLists[i];
      while (temp) {
        Node* toFree = temp;
        temp = temp->next;
        free(toFree);
      }
    }
    free(graph->adjLists);
    free(graph->visited);
    free(graph);
  }

  return 0;
}

File: code/zig/08-dfs_search.zig
const std = @import("std");
const print = std.debug.print;
const time = std.time;
const Instant = time.Instant;
const allocator = std.heap.page_allocator;
const Allocator = std.mem.Allocator;

const MAX_NODES: usize = 50_000; // Adjust as needed

/// Node structure for adjacency list
const Node = struct {
    vertex: usize,
    next: ?*Node,
};

/// Graph structure
const Graph = struct {
    numVertices: usize,
    adjLists: []?*Node, // Array of optional pointers to Nodes
    visited: []bool, // Array of visited flags

    /// Deinitialize the graph by freeing allocated memory
    fn deinit(self: *Graph, alloc: Allocator) void {
        // Free all nodes in adjacency lists
        for (self.adjLists) |nodePtr| {
            var node = nodePtr;
            while (node) |currentNode| {
                const nextNode = currentNode.next;
                alloc.destroy(currentNode);
                node = nextNode;
            }
        }
        alloc.free(self.adjLists);
        alloc.free(self.visited);
    }
};

/// Function to create a node
fn createNode(alloc: Allocator, v: usize) !*Node {
    const newNode = try alloc.create(Node);
    newNode.* = Node{
        .vertex = v,
        .next = null,
    };
    return newNode;
}

/// Function to create a graph
fn createGraph(alloc: Allocator, vertices: usize) !*Graph {
    const graph = try alloc.create(Graph);
    graph.numVertices = vertices;

    graph.adjLists = try alloc.alloc(?*Node, vertices);
    graph.visited = try alloc.alloc(bool, vertices);

    for (0..vertices) |i| {
        graph.adjLists[i] = null;
        graph.visited[i] = false;
    }

    return graph;
}

/// Function to add edge to an undirected graph
fn addEdge(alloc: Allocator, graph: *Graph, src: usize, dest: usize) !void {
    // Add edge from src to dest
    var newNode = try createNode(alloc, dest);
    newNode.next = graph.adjLists[src];
    graph.adjLists[src] = newNode;

    // Since the graph is undirected, add edge from dest to src also
    newNode = try createNode(alloc, src);
    newNode.next = graph.adjLists[dest];
    graph.adjLists[dest] = newNode;
}

/// DFS algorithm
fn dfs(graph: *Graph, vertex: usize) void {
    graph.visited[vertex] = true;

    var temp = graph.adjLists[vertex];
    while (temp) |node| {
        const adjVertex = node.vertex;

        if (!graph.visited[adjVertex]) {
            dfs(graph, adjVertex);
        }
        temp = node.next;
    }
}

/// Main function to demonstrate DFS and measure execution time.
pub fn main() !void {
    const N_values = [_]usize{ 1_000, 5_000, 10_000, 20_000, 50_000 };

    for (N_values) |N| {
        if (N > MAX_NODES) {
            print("N exceeds MAX_NODES limit.\n", .{});
            continue;
        }

        const graph = try createGraph(allocator, N);
        defer allocator.destroy(graph);
        defer graph.deinit(allocator);

        // Build a connected graph (e.g., a chain)
        for (0..N - 1) |i| {
            try addEdge(allocator, graph, i, i + 1);
        }

        // Measure start time
        const start = try Instant.now();

        // Perform DFS starting from vertex 0
        dfs(graph, 0);

        // Measure end time
        const end = try Instant.now();

        // Calculate the elapsed time in seconds
        const elapsed: f64 = @floatFromInt(end.since(start));
        const time_spent = elapsed / time.ns_per_s;

        // Print the result
        print("DFS with N = {d}\n", .{N});
        print("Time taken: {d} seconds\n\n", .{time_spent});
    }
}
